File: code/python/LinkedList/Simple/LinkedList.py
from utils import fix_index

from Node import Node
from StructBase import StructBase
from LinkedList.base import LinkedListBase


class LinkedList(StructBase, LinkedListBase):
    _top: Node = None

    def __init__(self):
        super()

    def __str__(self) -> str:
        str_ = "Top -> "
        str_ += " -> ".join([str(e) for e in self.__iter__()])
        str_ += " -> NULL\n"
        return str_

    def __iter__(self) -> "LinkedList":
        self._tmpNode = self._top
        return self

    def __next__(self) -> int:
        if self._tmpNode is None:
            raise StopIteration
        data = self._tmpNode
        self._tmpNode = data.link
        return data

    def __len__(self) -> int:
        return self.size

    def __getitem__(self, index: int) -> Node:
        if self.is_empty():
            return None

        index = fix_index(index, self.size)

        for jdx, elem in enumerate(self.__iter__()):
            if index == jdx:
                return elem

    def is_empty(self) -> bool:
        return self._top is None

    def is_full(self) -> bool:
        return False

    def insert(self, value: Node, index: int = 0):
        if self.is_empty() or 0 == index:
            if not self.is_empty():
                value.link = self._top
            self._top = value
            self.size += 1
            return

        index = fix_index(index, self.size)
        if index != self.size - 1:
            index -= 1

        prevNode = self.__getitem__(index)
        if prevNode is None:
            return
        value.link = prevNode.link
        prevNode.link = value
        self.size += 1

    def removeAt(self, index) -> Node:
        if self.is_empty():
            return None

        tmpNode = self._top
        if 0 == index:
            self._top = tmpNode.link
            tmpNode.link = None
            self.size -= 1
            return tmpNode

        tmpNode = self.__getitem__(index - 1)

        rmNode = tmpNode.link
        if rmNode is None:
            return None
        if rmNode.link is None:
            tmpNode.link = None
        else:
            tmpNode.link = rmNode.link
        rmNode.link = None
        self.size -= 1
        return rmNode

File: code/typescript/LinkedList/Simple/LinkedList.ts
import { Node } from "../../Node/Node";
import { fixIndex } from "../../utils/utils";
import StructBase from "../../StructBase/StructBase";
import LinkedListInterface from "../interface/LinkedListInterface";

export default class LinkedList
  extends StructBase
  implements LinkedListInterface
{
  private _top: Node | undefined;

  constructor() {
    super();
  }

  *[Symbol.iterator]() {
    let tmpNode = this._top;
    let auxNode = this._top;
    while (tmpNode !== undefined) {
      auxNode = tmpNode;
      tmpNode = tmpNode.link;
      yield auxNode;
    }
  }

  isEmpty(): boolean {
    return !this._top;
  }

  isFull(): boolean {
    return false;
  }

  toString(): string {
    let str = "Top -> ";
    if (this.isEmpty()) return str;
    str += [...this].map((elem) => elem.data).join(" -> ");
    str += this.isEmpty() ? "NULL" : " -> NULL";
    return str;
  }

  at(index: number): Node | undefined {
    if (this.isEmpty()) return undefined;

    index = fixIndex(index, this.size);

    let idx = 0;
    let auxNode = this._top;
    for (let tmpNode of this) {
      if (idx === index) {
        auxNode = tmpNode;
        break;
      }
      idx++;
    }
    return auxNode;
  }

  insert(value: Node, index: number = 0) {
    if (0 === index || this.isEmpty()) {
      if (!this.isEmpty()) value.link = this._top;
      this._top = value;
      this.size++;
      return;
    }

    index = fixIndex(index, this.size);
    if (index !== this.size - 1) index -= 1;

    let prevNode = this.at(index);
    if (!prevNode) return;
    value.link = prevNode.link;
    prevNode.link = value;
    this.size++;
  }

  push(value: Node) {
    this.insert(value, 0);
  }

  append(value: Node) {
    this.insert(value, -1);
  }

  removeAt(index: number): Node | undefined {
    if (this.isEmpty() || !this._top) return undefined;

    let tmpNode: Node | undefined = this._top;

    if (!tmpNode) return tmpNode;
    if (index === 0) {
      this._top = tmpNode.link;
      tmpNode.link = undefined;
      this.size--;
      return tmpNode;
    }

    tmpNode = this.at(index - 1);
    if (!tmpNode) return tmpNode;

    let rmNode = tmpNode.link;
    if (!rmNode) return rmNode;

    if (!rmNode.link) tmpNode.link = undefined;
    else tmpNode.link = rmNode.link;
    rmNode.link = undefined;
    this.size--;
    return rmNode;
  }

  removeFirst(): Node | undefined {
    return this.removeAt(0);
  }

  removeLast(): Node | undefined {
    return this.removeAt(-1);
  }
}
